sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/BasicWriteStatsTracker.scala

```diff
@@ -30,12 +32,13 @@ import org.apache.spark.sql.execution.metric.{SQLMetric, SQLMetrics}
 import org.apache.spark.util.SerializableConfiguration
 
 
+
 /**
  * Simple metrics collected during an instance of [[FileFormatDataWriter]].
  * These were first introduced in https://github.com/apache/spark/pull/18159 (SPARK-20703).
  */
 case class BasicWriteTaskStats(
-    numPartitions: Int,
+    partitions: Seq[InternalRow],
```

Comment:
This increases the data size we need to transfer between executors and the driver. Do we have a microbenchmark to verify the impact?

Reply:
A benchmark for latency or memory usage? I can try it ~

Reply:
end-to-end performance of an INSERT query, with a partitioned table with 1 or 2 or 3 partitioned columns.

Reply:
ok ~ busy with my own work, will give feedback later